### PR TITLE
COMP: Allow multi config builds

### PR DESCRIPTION
**Pull request by hjmjohnson**
ENH: Simplified multi-config generator wrapping
    
    Python wrapping requires a very rigid directory layout
    for the python package that is not compatible with the
    the default generated layout of the multi-config
    cmake generators (VS, XCode, Ninja Multi-Config).
    
    Reduce the number of local variables, or implicit
    pathing requirement enforcement, representing
    the same concepts. The current code prefers
    using consistent variable names at a more global
    scope to enforce package layout requirements.
    
    The multi-config and the standard cmake generated
    projects now consistently produce the same
    python package layout:

## PR Checklist
- [X] No [API changes](https://github.com/InsightSoftwareConsortium/ITK/blob/master/CONTRIBUTING.md#breaking-changes) were made (or the changes have been approved)
- [X] No [major design changes](https://github.com/InsightSoftwareConsortium/ITK/blob/master/CONTRIBUTING.md#design-changes) were made (or the changes have been approved)

### CODE REVIEW

**Comment by hjmjohnson** [comment]
@dzenanz Can you please test a windows python build?

**Comment by dzenanz** [comment]
I will kick off a build, then review.

File: Wrapping/CMakeLists.txt [hunk 1]
```diff
@@ -36,6 +36,39 @@ project(WrapITK
...
+  option(ALLOW_MULTICONFIG_WRAPPING "Allow (dangerous) multi-config wrapping, USE WITH CAUTION!" OFF)
+  mark_as_advanced(ALLOW_MULTICONFIG_WRAPPING)
+
+  if(ALLOW_MULTICONFIG_WRAPPING)
+    if(NOT CMAKE_BUILD_TYPE)
+      message(FATAL_ERROR "Wrapping requires explicit setting of CMAKE_BUILD_TYPE")
+    endif()
+    list(LENGTH CMAKE_CONFIGURATION_TYPES NUM_CMAKE_CONFIGURATION_TYPES)
+    if(NOT NUM_CMAKE_CONFIGURATION_TYPES EQUAL 1)
+      message(WARNING "CMake multi-config generators are not supported with ITK_WRAPPING:BOOL=ON.
```
Comment (dzenanz) [inline Wrapping/CMakeLists.txt:52]: I did not see this warning despite having CMAKE_BUILD_TYPE=Release and CMAKE_CONFIGURATION_TYPES=RelWithDebInfo.

Comment (hjmjohnson) [reply Wrapping/CMakeLists.txt:52]: This was disabled by default.  Perhaps it should be enabled by default?

Comment (dzenanz) [reply Wrapping/CMakeLists.txt:52]: If it leads to problems, it definitely should. I had no need to fiddle with CMAKE_BUILD_TYPE as it did not affect anything in multi-config generators. So it always had the default value.

Comment (hjmjohnson) [reply Wrapping/CMakeLists.txt:52]: This was more important earlier, then I found a way to make the process more clean.

I was on the fence about completely removing this.

If the building work, I propose that we just remove this logic.

**Comment by dzenanz** [comment]
This built and ran tests successfully!
```text
ctest -C RelWithDebInfo -j5 -R Python
...
100% tests passed, 0 tests failed out of 109
```

**Comment by hjmjohnson** [comment]
@dzenanz YEAH!  Thanks. 🙏   I'll remove the unused safety code.

```bash
#psuedocode for testing all
cmake -G "Ninja Multi-Config" ../ITK

ninja -f build-Debug.ninja
ctest -C Debug -j 55 -R Python

ninja -v remove_all_wrapped_libs
ninja -f build-Release.ninja
ctest -C Release -j 55 -R Python

ninja -v remove_all_wrapped_libs
ninja -f build-RelWithDebInfo.ninja
ctest -C RelWithDebInfo -j 55 -R Python
```

**Comment by hjmjohnson** [comment]
@thewtex This patchset implements a more elegant solution than what I proposed early this week during our discussion.

File: CMake/ITKModuleExternal.cmake [hunk 2]
```diff
@@ -35,21 +35,53 @@ if(NOT DEFINED CMAKE_FIND_PACKAGE_NO_PACKAGE_REGISTRY)
...
+    #
+    # https://cmake.org/cmake/help/v3.10/prop_tgt/LIBRARY_OUTPUT_DIRECTORY.html#prop_tgt:LIBRARY_OUTPUT_DIRECTORY
+    # Multi-configuration generators (VS, Xcode) append a per-configuration subdirectory
+    # to the specified directory unless a generator expression is used.
+    # Using an always true generator expression to disable multi-config standard behavior
+    #
+    # When wrapping, the multi-config generators can only be used in degraded state
+    # of allowing only a single element int the CMAKE_CONFIGURATION_TYPES and enforcing
+    # that CMAKE_BUILD_TYPE match that type (see Wrapping/CMakeLists.txt enforcement)
+    set(CMAKE_LIBRARY_OUTPUT_DIRECTORY "$<1:${ITK_PYTHON_PACKAGE_DIR}>" CACHE PATH "Shared library directory with generator override")
```
Comment (thewtex) [inline CMake/ITKModuleExternal.cmake:79]: Cool

**Review by thewtex — APPROVED** [review]
Thanks, Hans!

**Comment by thewtex** [comment]
The Windows CI tests are failing after this merge:

https://dev.azure.com/itkrobotwindowpython/ITK.Windows.Python/_build?definitionId=1&_a=summary

```
181: Loading ITKPyBase... done
181: Loading ITKCommon... Traceback (most recent call last):
181:   File "D:/a/1/s/Modules/Core/SpatialObjects/wrapping/test/SpatialObjectTest.py", line 24, in <module>
181:     SOType = itk.SpatialObject[dim]
181:   File "D:\a\1\s-build\Wrapping\Generators\Python\itk\support\itkLazy.py", line 78, in __getattribute__
181:     itkBase.itk_load_swig_module(module, namespace)
181:   File "D:\a\1\s-build\Wrapping\Generators\Python\itk\support\itkBase.py", line 102, in itk_load_swig_module
181:     itk_load_swig_module(dep, namespace)
181:   File "D:\a\1\s-build\Wrapping\Generators\Python\itk\support\itkBase.py", line 110, in itk_load_swig_module
181:     l_module = loader.load(swig_module_name)
181:   File "D:\a\1\s-build\Wrapping\Generators\Python\itk\support\itkBase.py", line 262, in load
181:     l_spec.loader.exec_module(l_module)  # pytype: disable=attribute-error
181:   File "<frozen importlib._bootstrap_external>", line 728, in exec_module
181:   File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
181:   File "D:\a\1\s-build\Wrapping\Generators\Python\itk\support\..\ITKCommonPython.py", line 13, in <module>
181:     from . import _ITKCommonPython
181: ImportError: DLL load failed: The specified module could not be found.
181: itkTestDriver: Process exited with return value: 1
55/55 Test #181: PythonSpatialObjectTest .............................***Failed    0.55 sec
Loading ITKPyBase... done
Loading ITKCommon... Traceback (most recent call last):
  File "D:/a/1/s/Modules/Core/SpatialObjects/wrapping/test/SpatialObjectTest.py", line 24, in <module>
    SOType = itk.SpatialObject[dim]
  File "D:\a\1\s-build\Wrapping\Generators\Python\itk\support\itkLazy.py", line 78, in __getattribute__
    itkBase.itk_load_swig_module(module, namespace)
  File "D:\a\1\s-build\Wrapping\Generators\Python\itk\support\itkBase.py", line 102, in itk_load_swig_module
    itk_load_swig_module(dep, namespace)
  File "D:\a\1\s-build\Wrapping\Generators\Python\itk\support\itkBase.py", line 110, in itk_load_swig_module
    l_module = loader.load(swig_module_name)
  File "D:\a\1\s-build\Wrapping\Generators\Python\itk\support\itkBase.py", line 262, in load
    l_spec.loader.exec_module(l_module)  # pytype: disable=attribute-error
  File "<frozen importlib._bootstrap_external>", line 728, in exec_module
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "D:\a\1\s-build\Wrapping\Generators\Python\itk\support\..\ITKCommonPython.py", line 13, in <module>
```

**Pull request merged**